feat: Add S3 implementation for object_store

yorik opened the pull request:
## What ❔

Add S3 implementation for object_store.
<!-- What are the changes this PR brings about? -->
<!-- Example: This PR adds a PR template to the repo. -->
<!-- (For bigger PRs adding more context is appreciated) -->

## Why ❔

To remove hard dependency on GCS, add possibility to use any S3-like storage for snapshots and provers.
<!-- Why are these changes done? What goal do they contribute to? What are the principles behind them? -->
<!-- The `Why` has to be clear to non-Matter Labs entities running their own ZK Chain -->
<!-- Example: PR templates ensure PR reviewers, observers, and future iterators are in context about the evolution of repos. -->

## Is this a breaking change?
- [ ] Yes
- [x] No

## Operational changes
<!-- Any config changes? Any new flags? Any changes to any scripts? -->
<!-- Please add anything that non-Matter Labs entities running their own ZK Chain may need to know -->

## Checklist

<!-- Check your PR fulfills the following items. -->
<!-- For draft PRs check the boxes as you complete them. -->

- [x] PR title corresponds to the body of PR (we generate changelog entries from PRs).
- [x] Tests for the changes have been added / updated.
- [x] Documentation comments have been added / updated.
- [x] Code has been formatted via `zkstack dev fmt` and `zkstack dev lint`.


ref ZKD-2415

slowli commented on core/lib/object_store/src/s3.rs:
```diff
@@ -0,0 +1,225 @@
+        formatter
+            .debug_struct("S3Store")
+            .field("bucket_prefix", &self.bucket_prefix)
+            // Skip `client` as its representation may contain sensitive info
```
Nit: Is `endpoint` potentially sensitive as well? Otherwise, it could contain useful debug info.

yorik replied:
Added `endpoint` here. It isn't sensitive.

slowli commented on core/lib/object_store/src/s3.rs:
```diff
@@ -0,0 +1,225 @@
+            .or_else(Region::new("auto"));
+        let mut sdk_config = Self::get_client_config(auth_mode).region(region_provider);
+        if let Some(endpoint) = endpoint.clone() {
+            tracing::info!(endpoint=%endpoint, "using S3 endpoint defined in storage config");
```
Nit: `endpoint=%endpoint` can be shortened to just `%endpoint`.

yorik replied:
Done

slowli commented on core/lib/object_store/src/s3.rs:
```diff
@@ -0,0 +1,225 @@
+                }
+            }
+            SdkError::ServiceError(http_response) => {
+                let code = http_response.raw().status().as_u16();
```
Nit: It looks like handling of this and the previous variant can be extracted to a function in order to avoid code duplication.

yorik replied:
I tried to do that, but unfortunately `http_response.raw()` returns private Response type, so I can't create a function which accept it. And `http_response` type is different in both cases.

slowli commented on core/lib/object_store/src/s3.rs:
```diff
@@ -0,0 +1,225 @@
+        self.client
+            .put_object()
+            .bucket(self.bucket_prefix.clone())
+            .key(&filename)
```
Nit: Since the API requires `impl Into<String>` here and `filename` is not used below, it would be more efficient to pass it by value: `key(filename)`.

yorik replied:
Done

slowli commented on core/lib/object_store/examples/object_store.rs:
```diff
@@ -0,0 +1,73 @@
+            .put::<SnapshotStorageLogsChunk>(key, &snapshot_chunk)
+            .await?;
+        let result = object_store
+            .get::<SnapshotStorageLogsChunk>(key.into())
```
Nit: Useless conversion for `key`. Ditto below.

yorik replied:
Done

slowli commented on core/Cargo.toml:
```diff
@@ -108,6 +108,9 @@ anyhow = "1"
 assert_matches = "1.5"
 async-trait = "0.1"
 async-recursion = "1"
+aws-config = { version = "1.1.7", features = ["behavior-version-latest"] }
+aws-runtime = "1.5.5"
+aws-sdk-s3 = "1.76.0"
```
Could you please check whether all default features are required for these crates to work? E.g., `aws-config` has the `sso` feature enabled by default bringing additional deps, which looks unused; at least the code compiles with

```toml
aws-config = { version = "1.1.7", default-features = false, features = ["behavior-version-latest"] }
```

yorik replied:
Perfect idea, removed 2 dependencies.

slowli reviewed (commented):
Overall, looks very nice 👍 I guess we have no real alternatives if we want to embed S3 support directly into the config / every binary using `ObjectStoreFactory`. (Otherwise, it could be implemented as a a custom `ObjectStore` in a new crate.)

yorik commented:
> Overall, looks very nice 👍 I guess we have no real alternatives if we want to embed S3 support directly into the config / every binary using `ObjectStoreFactory`. (Otherwise, it could be implemented as a a custom `ObjectStore` in a new crate.)

Thank you so much for the review! I've addressed all the comments, except of the error handling, see the explanation there.

I think we'd like to have direct support of S3 for easier migration to it were needed: the first will be snapshot-creator and EN. All the prover stack could follow later.

Could you please take another look on the PR?
